### ISSUE TITLE
`manual_unwrap_or_default` does not consider type changes and suggests wrong code

### ISSUE THREAD

**Issue by sdroege** [rust-lang/rust-clippy#12928]
### Summary

See code below

### Lint Name

manual_unwrap_or_default

### Reproducer

I tried this code:

```rust
fn main() {
    let x = Some((1, 2));
    let y = if let Some((a, _)) = x {
        a
    } else {
        0
    };
    
    println!("{y}");
}
```

I saw this happen:

```
warning: if let can be simplified with `.unwrap_or_default()`
 --> src/main.rs:3:13
  |
3 |       let y = if let Some((a, _)) = x {
  |  _____________^
4 | |         a
5 | |     } else {
6 | |         0
7 | |     };
  | |_____^ help: replace it with: `x.unwrap_or_default()`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#manual_unwrap_or_default
  = note: `#[warn(clippy::manual_unwrap_or_default)]` on by default
```

I expected to see this happen: [nothing]

### Version

```text
rustc 1.79.0 (129f3b996 2024-06-10)
binary: rustc
commit-hash: 129f3b9964af4d4a709d1383930ade12dfe7c081
commit-date: 2024-06-10
host: x86_64-unknown-linux-gnu
release: 1.79.0
LLVM version: 18.1.7
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by GuillaumeGomez** [comment]
Taking a look.
